### PR TITLE
Add new programming exercise for Swift

### PR DESCRIPTION
**Pull request by derLalla**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context

We want to have Swift exercises in Artemis, so that it also might be used in the iPraktikum.

### Description

This PR adds a template for Swift programming exercises but doesn't contain valid tests for the CI, yet.
A regex pattern for Swift package names was added, so that only unseparated names are allowed, such as "SwiftEx1" and not "Swift_Ex.1"

### Steps for Testing

1. Log in to Artemis
2. Navigate to Course Administration
3. Create a new Programming Exercise, select Swift in the dropdown menu and enter a package name
4. Verify it was successful and start the exercise
4. Click on open in Online Editor 
5. See a working template for BubbleSort and MergeSort as a Swift code exercise which package's name is the one defined in step 3.
6. Verify that the ACE module could be loaded and no 404 error appears in the console.

Note: The new programming languages are only supported with Bamboo, so make sure to use TS1 or TS3 when testing.

### Test Coverage

The test coverage is almost staying the same, because only existing files were edited, which already have parameterized tests like the Programming Language Enum.

### Screenshots

image

image

image

image

### CODE REVIEW

**Comment by stefanwaldhauser** [comment]
I get this error when I access the code editor: 
image

**Review by fde312 — CHANGES_REQUESTED** [review]
I am getting a 404 error when opening the code editor with a swift exercise, same issue as found by @stefanwaldhauser :
image

Furthermore, the template has not tests and when I retrigger the build I get build failed (tried it out for two exercises):
image

image

**Review by MaisiKoleni — COMMENTED** [review]
Solution passes the tests, but the template/submissions only report that the build failed due to failing tests (if I interpret the logs correctly). Is test reporting in the scope of this PR? We should report failed tests such that Artemis recognizes them.

I can also observe the problems the other two reported.

**Review by kloessst — CHANGES_REQUESTED** [review]
image

For me the result of the initial template is not correctly displayed. I think this should work before we merge this PR. I'll approve, if this is actually expected.

**Comment by derLalla** [comment]
> I get this error when I access the code editor:
> image
> I think the ace-editor tries to load it.

@stefanwaldhauser and @fde312 this is also happening when opening a C exercise. I guess that this was also never noticed. 
I have found that the ACE editor is trying to load libraries only for some programming languages. (No external loading in Java files). 

The mentioned lib "mode-swift.js" and for C "mode-makefile.js" can both be found here: https://kibotics.org/static/ace-editor/src-min/ 

But what is the correct approach?

**Comment by derLalla** [comment]
> image
> 
> For me the result of the initial template is not correctly displayed. I think this should work before we merge this PR. I'll approve, if this is actually expected.

I've fixed the issue

**Comment by derLalla** [comment]
> > image
> > For me the result of the initial template is not correctly displayed. I think this should work before we merge this PR. I'll approve, if this is actually expected.
> 

I've fixed the issue

**Comment by derLalla** [comment]
> Solution passes the tests, but the template/submissions only report that the build failed due to failing tests (if I interpret the logs correctly). Is test reporting in the scope of this PR? We should report failed tests such that Artemis recognizes them.
> 
> I can also observe the problems the other two reported.

I've fixed the issue

File: src/main/resources/templates/swift/exercise/Sources/${packageNameFolder}Lib/BubbleSort.swift
Comment (fde312) [inline src/main/resources/templates/swift/exercise/Sources/${packageNameFolder}Lib/BubbleSort.swift:5]:
```suggestion
	public func performSort(_ input: input [Date]) {
```

**Review by fde312 — COMMENTED** [review]
For me now the template works:
image
And I was also able to create a rated submission:
image

@derLalla I am just wondering, because I solved MergeSort with the code for BubbleSort 😄 
You also have some codacy code style issues, would be nice if you can take a look into those.

File: src/main/webapp/app/exercises/text/participate/text-result/text-result-block.ts
Comment (fde312) [inline src/main/webapp/app/exercises/text/participate/text-result/text-result-block.ts:32]: Please revert this change. It is intended to be `== undefined`. Codacy will not report this anymore in the latest develop
```suggestion
        if (!this.feedback || this.feedback.credits == undefined || Feedback.isEmpty(this.feedback)) {
```

Comment (derLalla) [reply src/main/webapp/app/exercises/text/participate/text-result/text-result-block.ts:32]: But my local setup was always complaining that this is an error and wouldn't start the client

**Comment by derLalla** [comment]
> For me now the template works. Good job! 🚀
> image
> And I was also able to create a rated submission:
> image
> 
> @derLalla I am just wondering, because I solved MergeSort with the code for BubbleSort 😄
> You also have some codacy code style issues, would be nice if you can take a look into those.

Yes, i guess it would also be possible in the Java exercise to solve the MergeSort by using the BubbleSort algorithm. It only checks if the correct interface is used. 
The Swift test cases don't support the class and interface checks like the Java test cases do.

File: src/main/webapp/app/exercises/text/participate/text-result/text-result-block.ts
Comment (derLalla) [reply src/main/webapp/app/exercises/text/participate/text-result/text-result-block.ts]: fixed

Comment (fde312) [reply src/main/webapp/app/exercises/text/participate/text-result/text-result-block.ts]: Yes, this was due to prettier complaining about your code style. This is fixed with the latest develop merge. :)

**Review by stefanwaldhauser — COMMENTED** [review]
Everything worked for me now :) Gj @derLalla

**Review by stefanwaldhauser — DISMISSED** [review]
Worked as expected on TS3 nice work :)

**Review by fde312 — APPROVED** [review]
Thanks for incorporating the feedback! 👍

**Review by krusche — CHANGES_REQUESTED** [review]
Nice contribution, the code looks generally very good. I only found some minor issues:
* Please fill out all section in the PR template (Screenshot + Test Coverage)
* There is a warning about static final, please fix it as well
* Double check the README so it is consistent (e.g. the plantUML config) with other programming languages

**Review by stefanwaldhauser — APPROVED** [review]
Reapprove

File: src/main/resources/templates/swift/exercise/Sources/${packageNameFolder}Lib/Client.swift
Comment (FrankeLukas) [inline src/main/resources/templates/swift/exercise/Sources/${packageNameFolder}Lib/Client.swift:5]: Minor nit: I assume this wants the same 4-space indentation like the following `TODO`s.

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java
Comment (FrankeLukas) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java:338]: It probably isn't useful to match on the regular package name Regex first if we're dealing with a Swift exercise. What do you think about changing this to something like
```
Matcher packageNameMatcher = null;
if (programmingExercise.getProgrammingLanguage() == ProgrammingLanguage.SWIFT) {
    packageNameMatcher = packageNamePatternForSwift.matcher(programmingExercise.getPackageName());
} else {
    packageNamePattern.matcher(programmingExercise.getPackageName());
}
```
?

**Review by FrankeLukas — DISMISSED** [review]
Creating and submitting a Swift programming exercise worked fine.

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java
Comment (kloessst) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java:338]: Good suggestion!

**Review by kloessst — APPROVED** [review]
Code looks good and works for me. I think @FrankeLukas is right about the package regex. Otherwise good job!

**Comment by derLalla** [comment]
> Nice contribution, the code looks generally very good. I only found some minor issues:
> 
> * Please fill out all section in the PR template (Screenshot + Test Coverage)
> * There is a warning about static final, please fix it as well
> * Double check the README so it is consistent (e.g. the plantUML config) with other programming languages

thanks, I've fixed it. :)

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java
Comment (derLalla) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java:338]: @FrankeLukas your suggestion is basically the same but more lines of code, right? In case we don't have a Swift Exercise, the packageNameMatcher gets assigned the packageNamePattern in the else part. To save this else part I am assigning it always instead of first setting it to null.

File: src/main/resources/templates/swift/exercise/Sources/${packageNameFolder}Lib/Client.swift
Comment (derLalla) [reply src/main/resources/templates/swift/exercise/Sources/${packageNameFolder}Lib/Client.swift:5]: keen eye!

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java
Comment (derLalla) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java:338]: but I guess you're right. I've fixed it :)

**Review by fde312 — DISMISSED** [review]
re-approve

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java
Comment (MaisiKoleni) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java:110]:
```suggestion
    private static final String packageNameRegex = "^(?!.*(?:\\.|^)(?:abstract|continue|for|new|switch|assert|default|if|package|synchronized|boolean|do|goto|private|this|break|double|implements|protected|throw|byte|else|import|public|throws|case|enum|instanceof|return|transient|catch|extends|int|short|try|char|final|interface|static|void|class|finally|long|strictfp|volatile|const|float|native|super|while|_|true|false|null)(?:\\.|$))[A-Z_a-z][0-9A-Z_a-z]*(?:\\.[A-Z_a-z][0-9A-Z_a-z]*)*$";

    private static final String packageNameRegexForSwift = "^(?!.*(?:\\.|^)(?:abstract|continue|for|new|switch|assert|default|if|package|synchronized|boolean|do|goto|private|this|break|double|implements|protected|throw|byte|else|import|public|throws|case|enum|instanceof|return|transient|catch|extends|int|short|try|char|final|interface|static|void|class|finally|long|strictfp|volatile|const|float|native|super|while|_|true|false|null)(?:\\.|$))[A-Za-z][0-9A-Za-z]*$";
```

Comment (MaisiKoleni) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java:110]: This does not look like the correct Swift package pattern, I don't know Swift, but this looks odd to me. I would recommend going through https://docs.swift.org/swift-book/ReferenceManual/LexicalStructure.html#ID413 or similar and exchange the Java keywords by the Swift ones (same on the client side). It is very likely other things need to be adjusted as well. (E.g. if Swift does not support `.` between the package parts, the first part of the regex needs to be adjusted as well.)

**Review by MaisiKoleni — CHANGES_REQUESTED** [review]
See comment on the package name regex, this should be fixed or at least commented/explained (or add a `TODO`).

**Comment by krusche** [comment]
I agree to the changes requested by @MaisiKoleni, thanks for the review.
Would be great if you could fix those issues, @derLalla

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java
Comment (derLalla) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java:110]: `.` are not allowed and not getting captured

Comment (derLalla) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java:110]: But actually it wouldn't cause any problems, as the package name gets appended by a suffix "App" and "Lib", so that the test cases are able to compile. 
Nevertheless I have adapted the key words.

**Review by stefanwaldhauser — DISMISSED** [review]
re-approve

**Review by stefanwaldhauser — DISMISSED** [review]
re-approve

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java
Comment (MaisiKoleni) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseResource.java:114]:
```suggestion
    private static final String packageNameRegexForSwift = "^(?!(?:associatedtype|class|deinit|enum|extension|fileprivate|func|import|init|inout|internal|let|open|operator|private|protocol|public|rethrows|static|struct|subscript|typealias|var|break|case|continue|default|defer|do|else|fallthrough|for|guard|if|in|repeat|return|switch|where|while|as|Any|catch|false|is|nil|super|self|Self|throw|throws|true|try|_)$)[A-Za-z][0-9A-Za-z]*$";
```
I think those parts where still left from the `.` separator. If it is not allowed and just one identifier is used for Swift exercises, this should be the right choice.

Currently, `somefunc` would not be allowed (`some` matching the `.*`), but I would assume it is.

File: src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.ts
Comment (MaisiKoleni) [inline src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.ts:63]:
```suggestion
        '^(?!(?:associatedtype|class|deinit|enum|extension|fileprivate|func|import|init|inout|internal|let|open|operator|private|protocol|public|rethrows|static|struct|subscript|typealias|var|break|case|continue|default|defer|do|else|fallthrough|for|guard|if|in|repeat|return|switch|where|while|as|Any|catch|false|is|nil|super|self|Self|throw|throws|true|try|_)$)[A-Za-z][0-9A-Za-z]*$';
```

**Review by MaisiKoleni — CHANGES_REQUESTED** [review]
I just have a small suggestion (please also check that it is correct), but code looks good to me.

**Review by tiffca — DISMISSED** [review]
Tested successfully on test server 1. Confirmed creating/participating in a swift exercise works well using the provided solution. The package regex works as described, and no errors seen in the console.

**Review by kloessst — APPROVED** [review]
- Creating/Importing a Swift programming exercise works fine
- Template participation: 0%; Solution participation: 100%
- Participating as a student works
- Ace editor does not attempt to load Swift dependencies
- Code looks good to me

--> Approve

**Review by MaisiKoleni — APPROVED** [review]
Code looks good to me now.

**Review by stefanwaldhauser — APPROVED** [review]
re-approve

**Pull request merged**
